gbpservice/neutron/nsf/configurator/agents/vpn.py

```diff
@@ -0,0 +1,617 @@
+        self.agent.update_service_status(
+            context,
+            vpnsvc,
+            const.STATE_ERROR)
+        raise exc.ResourceErrorState(name='vpn_service', id=vpnsvc['id'],
+                                     message=message)
+
+    def _active_state(self, context, vpnsvc):
+        self.agent.update_service_status(
+            context,
+            vpnsvc,
+            const.STATE_ACTIVE)
+
+    def _get_local_cidr(self, vpn_svc):
+        svc_desc = vpn_svc['description']
+        tokens = svc_desc.split(';')
+        local_cidr = tokens[1].split('=')[1]
+        return local_cidr
+
+    def validate(self, context, vpnsvc):
+        lcidr = self._get_local_cidr(vpnsvc)
+        """
+        Get the vpn services for this tenant
+        Check for overlapping lcidr - not allowed
+        """
+        filters = {'tenant_id': [context.tenant_id]}
+        t_vpnsvcs = self.agent.get_vpn_services(
+            context, filters=filters)
+        vpnsvc.pop("status", None)
+        for svc in t_vpnsvcs:
+            del svc['status']
+        if vpnsvc in t_vpnsvcs:
+            t_vpnsvcs.remove(vpnsvc)
+        for svc in t_vpnsvcs:
+            t_lcidr = self._get_local_cidr(svc)
+            if t_lcidr == lcidr:
+                msg = ("Local cidr %s conflicts with existing vpnservice %s"
+                       % (lcidr, svc['id']))
+                LOG.error(msg)
+                self._error_state(
+                    context,
+                    vpnsvc, msg)
+        self._active_state(context, vpnsvc)
+
+
+class VpnaasRpcSender(object):
+    """ RPC APIs to VPNaaS Plugin.
+    """
+    RPC_API_VERSION = '1.0'
+    target = messaging.Target(version=RPC_API_VERSION)
+
+    def __init__(self, topic, context):
+        self.context = context
+
+    def get_vpn_services(self, context, ids, filters):
+        """Get list of vpnservices on this host.
+        """
+        return self.call(
+            context,
+            self.make_msg('get_vpn_services', ids=ids, filters=filters))
+
+    def get_vpn_servicecontext(self, context, svctype,  filters):
+        """Get list of vpnservice context on this host.
+           For IPSEC connections :
+                List of vpnservices -->
+                lIst of ipsec connections -->
+                ike policy & ipsec policy
+        """
+        return self.call(
+            context,
+            self.make_msg(
+                'get_vpn_servicecontext',
+                svctype=svctype,  filters=filters),
+            version=self.API_VERSION)
+
+    def get_ipsec_conns(self, context, filters):
+        """
+        Get list of ipsec conns with filters
+        specified.
+        """
+        return self.call(
+            context,
+            self.make_msg(
+                'get_ipsec_conns',
+                filters=filters),
+            version=self.API_VERSION)
+
+    def update_status(self, context, status):
+        """Update local status.
```

Comment:
Need to enqueue to notification queue. Check firewall agent for example data format and enqueue.